### ISSUE TITLE
Add generic comparison operators

### ISSUE THREAD

**Issue by maz** [JHU-PL-Lab/big-bang#12]
@leafac mentioned in #10 that, right now, the comparison operators in big bang desugar to calling the tiny bang integer operation builtins, and therefore cannot work for other data types. To fix this, one option is to have the comparison operators desugar to a function, which can be onioned, so that it works on both ints and chars and other types. However, at the moment, it looks like we don't have a standard prelude anywhere, so there's no place to put these functions. So, how should we solve this problem? We could have a little bang "prelude" which is just little bang code which is pasted at the top of all little bang code which big bang generates.

**Comment by maz** [comment]
Tagging @rpglover64 @zepalmer

**Comment by zepalmer** [comment]
I agree; the correct approach is to have a preamble of some kind which is attached during desugaring.  @leafac and I have discussed this, but it just hasn't happened yet.  We agree with the approach you're proposing, if I understand correctly.

**Comment by leafac** [comment]
I agree with the approach of having a preamble in the translated Tiny Bang programs consisting of useful helpers (e.g. the comparison function mentioned in this thread).

But I don't think it is a prerequisite to do the comparison function itself. We can embed it everywhere it's used - the same way we're doing with the other helper functions (e.g. the Y Combinator used in function abstractions (named functions)).

**Comment by zepalmer** [comment]
We can embed the comparison function everywhere it's used _for now_, but it will be necessary to use a global preamble before we can do things like lexically-scoped operator overloading.  Perhaps this preamble idea should be it's own issue report?

**Comment by leafac** [comment]
@zepalmer 5 seconds ago I started working on getting rid of that `docs/roadmap.org` file. So an issue is going to be created for this in a few minutes :smile:

**Comment by leafac** [comment]
I'm adding one snippet @zepalmer wrote when we first discussed this:

```
int=
int+

⊕ x y z ...

e1 =? e2

equal? (`1 (e1) & `2 (e2))

equal? =
    (fun `1 x:int * `2 y:int -> int= x y)
  & (fun `1 x:char * `2 y:char -> char= x y)


g = fun p1 \ { p1 = `A p2, p2 = () } -> ...
f = fun p0 \ { p0 = x : int } -> { r = x }
a = 1
b = 'z'
c = a & b
z = f c

```

**Comment by leafac** [comment]
Would it be feasible to have comparison defined in objects/records? The default implementation would be to check if the fields are the same and recurse on the values. The programmer could overload that and implement more clever comparison methods for their data types.

From what I remember, this is impossible in Big Bang because we don't have a way to list all the keys of an object/record. That kind of introspection/reflection would break the theory.

But I want to double check, because I feel this would be really useful.

**Comment by zepalmer** [comment]
We explored something similar to this several years ago and discovered it to be a giant pain.  It started with our exploration of a fully polymorphic equality operator; that is, we wanted to be able to check if ``A 5 & `B 3 == `A 5 & `B 3`.  This turned out to be a rabbit hole.  The first concerning question, for instance, is whether ``A 5 & `B 3 == `A 5 & `B 3 & `C 2`.  If we say "yes", we could be unpleasantly surprised; if we say "no", we could be violating intuitions about duck typing.  The problem is further complicated by ``A 5 & `B 3 == `B 3 & `C 2`.

Your proposed semantics above sound like testing for exact structural equality; that is, you would say "no" to both of the questions above.  But we can consider, for instance, that ``C` may actually be ``SanitizedString` or some other irrelevant tagging structure that is expected to transparently follow the data around.

Eventually, we boxed the issue.  If you have a custom data type, it's really up to your problem domain whether two instances of that type are "equal".  The structural equality that you propose might be useful (in the same way that OCaml's structural equality is useful -- that is, up to a point), but I'd prefer to define custom equality operations for each of my data types (possibly through the use of a code preprocessor or metaprogramming system).

**Comment by leafac** [comment]
@zepalmer I understand your point.

In many cases, I want to define my custom equality operator. Your ``SanitizedString` being a great example of that.

I still there's value to provide exact structural equality. It's a reasonable default for many cases. Maybe we could provide that as a library of some sort in the standard library?

**Comment by zepalmer** [comment]
I agree that, for many cases, it's a reasonable default.  It would, however, require a built-in operation to be specially defined; as you observe, it's not possible to encode that behavior.  Also, it's important to note with this very broad structural equality operator that there are no two arguments you can give it which produce a type error.  It gets kind of complicated in there, so I'd definitely call it a low priority.

**Comment by leafac** [comment]
@zepalmer Agreed.

Thanks for taking the time to explain this to me :smile:
